packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.kt

```diff
@@ -70,38 +81,91 @@ class FileDownloader(
-  constructor(filesDirectory: File, easClientID: String, configuration: UpdatesConfiguration, logger: UpdatesLogger, client: OkHttpClient) : this(filesDirectory, easClientID, configuration, logger) {
+  constructor(filesDirectory: File, easClientID: String, configuration: UpdatesConfiguration, logger: UpdatesLogger, client: OkHttpClient) : this(filesDirectory, easClientID, configuration, logger, null) {
     this.client = client
   }
 
   data class FileDownloadResult(val file: File, val hash: ByteArray)
   data class AssetDownloadResult(val assetEntity: AssetEntity, val isNew: Boolean)
 
   private suspend fun downloadAssetAndVerifyHashAndWriteToPath(
+    asset: AssetEntity,
     request: Request,
     expectedBase64URLEncodedSHA256Hash: String?,
     destination: File,
-    progressListener: FileDownloadProgressListener? = null
+    updatesDirectory: File,
+    progressListener: FileDownloadProgressListener? = null,
+    allowPatch: Boolean = true
   ): FileDownloadResult {
     try {
       val response = downloadData(request, progressListener)
 
-      if (!response.isSuccessful) {
-        val message = "Asset download request not successful"
-        val cause = IOException(response.body?.string() ?: "Unknown error")
-        logger.error(message, cause, UpdatesErrorCode.AssetsFailedToLoad)
-        throw IOException(message, cause)
-      }
+      response.use { resp ->
+        if (!resp.isSuccessful) {
+          val message = "Asset download request not successful"
+          val cause = IOException(resp.body?.string() ?: "Unknown error")
+          logger.error(message, cause, UpdatesErrorCode.AssetsFailedToLoad)
+          throw IOException(message, cause)
+        }
 
-      try {
-        response.body!!.byteStream().use { inputStream ->
-          val hash = UpdatesUtils.verifySHA256AndWriteToFile(inputStream, destination, expectedBase64URLEncodedSHA256Hash)
-          return FileDownloadResult(destination, hash)
+        val responseBody = resp.body
+          ?: throw IOException("Asset download response from ${request.url} had no body")
+
+        val isPatch = isPatchResponse(responseBody)
+        val hash = if (allowPatch && isPatch) {
+          try {
+            applyPatchAndVerify(
+              asset,
+              request,
+              responseBody,
+              destination,
+              updatesDirectory,
+              expectedBase64URLEncodedSHA256Hash
+            )
+          } catch (_: Exception) {
+            logger.warn(
+              "Hermes diff application failed for asset ${asset.key}; retrying with full asset download",
+              UpdatesErrorCode.AssetsFailedToLoad,
+              request.header("Expo-Requested-Update-ID"),
+              asset.key
+            )
+
+            responseBody.close()
+            val fallbackRequest = request.newBuilder()
+              .headers(request.headers)
+              .removeHeader("Accept")
+              .header("Accept", "application/javascript")
```

Comment:
I think this fallback request needs to still respect overrides in `configuration.requestHeaders`? for example, if I put `accept` in there, it'd override the accept added in this PR. Maybe there's a way we can refactor the request creation code or header code to share between this and `createRequestForAsset`? Or just call `createRequestForAsset` again with a flag that doesn't add the accept header?